virtualbox/vbox-export-snapshots.py

```diff
@@ -32,56 +32,217 @@
+            raise TimeoutError(f"VM did not start within the timeout period {timeout}s.")
+        else:
+            print("VM is already running.")
+            return
+    except Exception as e:
+        print(f"Error checking VM state: {e}")
+    raise Exception(f"Could not ensure '{machine_guid}' running")
+
+def ensure_vm_shutdown(machine_guid):
+    """Checks if the VM is running and shuts it down if it is."""
+    try:
+        vm_state = get_vm_state(machine_guid)
+        if vm_state != "poweroff":
+            print(f"VM {machine_guid} is not powered off. Shutting down VM...")
+            run_vboxmanage(["controlvm", machine_guid, "poweroff"]) 
+
+            # Wait for VM to shut down (up to 1 minute)
+            timeout = 60  # seconds
+            check_interval = 5  # seconds
+            start_time = time.time()
+            while time.time() - start_time < timeout:
+                vm_state = get_vm_state(machine_guid)
+                if vm_state == "poweroff":
+                    print(f"VM {machine_guid} is shut down (status: {vm_state}).")
+                    time.sleep(5) # wait a bit to be careful and avoid any weird races
+                    return
+                time.sleep(check_interval)
+            print("Timeout waiting for VM to shut down. Exiting...")
+            raise TimeoutError("VM did not shut down within the timeout period.")
+        else:
+            print(f"VM {machine_guid} is already shut down (state: {vm_state}).")
+            return
+    except Exception as e:
+        print(f"Error checking VM state: {e}")
+    raise Exception(f"Could not ensure '{machine_guid}' shutdown")
+
+def ensure_hostonlyif_exists():
+    """Gets the name of, or creates a new hostonlyif"""
+    try:
+        # Find existing hostonlyif
+        hostonlyifs_output = run_vboxmanage(["list", "hostonlyifs"])
+        for line in hostonlyifs_output.splitlines():
+            if line.startswith("Name:"):
+                hostonlyif_name = line.split(":")[1].strip()
+                print(f"Found existing hostonlyif {hostonlyif_name}")
+                return
+        
+        # No host-only interface found, create one
+        print("No host-only interface found. Creating one...")
+        run_vboxmanage(["hostonlyif", "create"])  # Create a host-only interface
+        hostonlyifs_output = run_vboxmanage(["list", "hostonlyifs"])  # Get the updated list
+        for line in hostonlyifs_output.splitlines():
+            if line.startswith("Name:"):
+                hostonlyif_name = line.split(":")[1].strip()
+                print(f"Created hostonlyif {hostonlyif_name}")
+                return
+        print("Failed to create new hostonlyif. Exiting...")
+        raise Exception("Failed to create new hostonlyif.")
+    except Exception as e:
+        print(f"Error getting host-only interface name: {e}")
+    raise Exception("Failed to verify host-only interface exists")
```

Comment:
The command will silently do nothing if there is no host only adapter. It would be caught by the check I added in the change adapter function, but I chose to ensure the VM state is as we expect. If I did not do this there would need to be a manual step to ensure a host only adapter exists which seemed not necessary, why not automate it.